nixos/modules/services/web-apps/coder.nix

```diff
@@ -0,0 +1,172 @@
+        '';
+        default = "/var/lib/coder";
+      };
+
+      listenAddress = mkOption {
+        type = types.str;
+        description = lib.mdDoc ''
+          Listen address.
+        '';
+        default = "127.0.0.1:3000";
+      };
+
+      accessUrl = mkOption {
+        type = types.nullOr types.str;
+        description = lib.mdDoc ''
+          Access URL should be a external IP address or domain with DNS records pointing to Coder.
+        '';
+        default = null;
+        example = "https://coder.example.com";
+      };
+
+      wildcardAccessUrl = mkOption {
+        type = types.nullOr types.str;
+        description = lib.mdDoc ''
+          If you are providing TLS certificates directly to the Coder server, you must use a single certificate for the root and wildcard domains.
+        '';
+        default = null;
+        example = "*.coder.example.com";
+      };
+
+      postgresqlUrl = mkOption {
+        type = types.nullOr types.str;
+        description = lib.mdDoc ''
+          Coder uses a PostgreSQL database to store users, workspace metadata, and other deployment information.
+        '';
+        default = "user=coder database=coder host=/run/postgresql sslmode=disable";
+        example = "user=coder database=coder host=/run/postgresql sslmode=disable";
+      };
+
+      tlsCert = mkOption {
+        type = types.nullOr types.path;
+        description = lib.mdDoc ''
+          The path to the TLS certificate.
+        '';
+        default = null;
+      };
+
+      tlsKey = mkOption {
+        type = types.nullOr types.path;
+        description = lib.mdDoc ''
+          The path to the TLS key.
+        '';
+        default = null;
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.services.coder = {
+      description = "Coder - Self-hosted developer workspaces on your infra";
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+
+      environment = {
+        CODER_ACCESS_URL = cfg.accessUrl;
+        CODER_WILDCARD_ACCESS_URL = cfg.wildcardAccessUrl;
+        CODER_PG_CONNECTION_URL = cfg.postgresqlUrl;
+        CODER_ADDRESS = cfg.listenAddress;
+        CODER_TLS_ENABLE = optionalString (cfg.tlsCert != null) "1";
+        CODER_TLS_CERT_FILE = cfg.tlsCert;
+        CODER_TLS_KEY_FILE = cfg.tlsKey;
+      };
+
+      serviceConfig = {
+        ProtectSystem = "full";
+        PrivateTmp = "yes";
+        PrivateDevices = "yes";
+        SecureBits = "keep-caps";
+        AmbientCapabilities = "CAP_IPC_LOCK CAP_NET_BIND_SERVICE";
+        CacheDirectory = "coder";
+        CapabilityBoundingSet = "CAP_SYSLOG CAP_IPC_LOCK CAP_NET_BIND_SERVICE";
+        KillSignal = "SIGINT";
+        KillMode = "mixed";
+        NoNewPrivileges = "yes";
+        Restart = "on-failure";
+        ExecStart = "${cfg.package}/bin/coder server";
+        User = cfg.user;
+        Group = cfg.group;
+      };
+    };
+
+    services.postgresql = {
```

Comment:
There should be a way to configure this

Reply:
Like to an option to configure the postgres user and database name?

Reply:
or one `preconfigurePostgres`?

Reply:
@SuperSandro2000 it is great if module authors want to add this, but personally I don't consider it a requisite to merge. If someone wants the functionality later they can add it...

That being said, @shyim is a superstar so they probably are willing to add the functionality. Something like this:

```
options.coder.database = {
  name = mkOption {
    type = types.str;
    default = "coder";
    description = ...;
  };

  host = mkOption {
    type = types.str;
    default = "localhost";
    description = ...;
  };

  user = mkOption {
    type = types.str;
    default = "coder";
    description = ...;
  };

  createLocally = mkOption {
    type = types.bool;
    default = true;
    description = ... whether to create the database and database user locally ...
  };
};

config.assertions = [ { assertion = cfg.database.createLocally -> cfg.database.host = "localhost"; message = ...; } ];
```

@shyim - take a look at [search.nixos.org/options](https://search.nixos.org/options?channel=22.11&show=services.redmine.database.createLocally&from=0&size=50&sort=relevance&type=packages&query=createLocally) for a large number of examples.

Reply:
its done